// @amazon-connect/core/lib-esm/context/module-context.js
import { ConnectLogger } from "../logging";
import { ConnectMetricRecorder, } from "../metric";
import { createModuleProxy } from "../proxy";
export class ModuleContext {
    constructor(engineContext, moduleNamespace) {
        this.engineContext = engineContext;
        this.moduleNamespace = moduleNamespace;
    }
    get proxy() {
        if (!this.moduleProxy) {
            const proxy = this.engineContext.getProvider().getProxy();
            const moduleNamespace = this.moduleNamespace;
            this.moduleProxy = createModuleProxy(proxy, moduleNamespace);
        }
        return this.moduleProxy;
    }
    getProvider() {
        return this.engineContext.getProvider();
    }
    createLogger(params) {
        if (typeof params === "object") {
            return new ConnectLogger(Object.assign(Object.assign({}, params), { provider: () => this.engineContext.getProvider() }));
        }
        else {
            return new ConnectLogger({
                source: params,
                provider: () => this.engineContext.getProvider(),
            });
        }
    }
    createMetricRecorder(params) {
        if (typeof params === "object") {
            return new ConnectMetricRecorder(Object.assign(Object.assign({}, params), { provider: () => this.engineContext.getProvider() }));
        }
        else {
            return new ConnectMetricRecorder({
                namespace: params,
                provider: () => this.engineContext.getProvider(),
            });
        }
    }
}
