// @amazon-connect/core/lib/utility/emitter/emitter-base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmitterBase = void 0;
const logging_1 = require("../../logging");
class EmitterBase {
    constructor({ provider, loggerKey }) {
        this.events = new Map();
        this.logger = new logging_1.ConnectLogger({
            provider,
            source: "emitter",
            mixin: () => ({
                emitterLoggerKey: loggerKey,
            }),
        });
    }
    on(parameter, handler) {
        const set = this.events.get(parameter);
        if (set)
            set.add(handler);
        else
            this.events.set(parameter, new Set([handler]));
    }
    off(parameter, handler) {
        const set = this.events.get(parameter);
        if (set) {
            set.delete(handler);
            if (set.size < 1)
                this.events.delete(parameter);
        }
    }
    getHandlers(parameter) {
        var _a;
        return Array.from((_a = this.events.get(parameter)) !== null && _a !== void 0 ? _a : []);
    }
}
exports.EmitterBase = EmitterBase;
