// @amazon-connect/core/lib/request/request-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestManager = void 0;
const logging_1 = require("../logging");
const request_handler_factory_1 = require("./request-handler-factory");
class RequestManager {
    constructor(provider) {
        this.requestMap = new Map();
        this.logger = new logging_1.ConnectLogger({
            provider,
            source: "core.requestManager",
        });
    }
    processRequest(request) {
        const { requestId } = request;
        return (0, request_handler_factory_1.createRequestHandler)(request, (handler) => this.requestMap.set(requestId, handler), ({ request, timeoutMs }) => this.handleTimeout(request, timeoutMs));
    }
    processResponse(response) {
        const { requestId } = response;
        const handler = this.requestMap.get(requestId);
        if (!handler) {
            // The proxy is implemented such that this should never happen
            this.logger.error("Returned a response message with no handler", {
                message: response,
            });
            return;
        }
        handler(response);
        this.requestMap.delete(requestId);
    }
    handleTimeout(request, timeoutMs) {
        const { requestId, namespace, command } = request;
        this.requestMap.delete(requestId);
        this.logger.error("Client request timeout", {
            requestId,
            namespace,
            command,
            timeoutMs,
        });
    }
}
exports.RequestManager = RequestManager;
