// @amazon-connect/core/lib/messaging/subscription/subscription-handler-id-map.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubscriptionHandlerIdMap = void 0;
const utility_1 = require("../../utility");
class SubscriptionHandlerIdMap {
    constructor() {
        this.idsByHandler = new Map();
        this.handlersById = new Map();
    }
    add(handler) {
        const existingId = this.idsByHandler.get(handler);
        if (existingId) {
            return { handlerId: existingId };
        }
        const handlerId = (0, utility_1.generateUUID)();
        this.idsByHandler.set(handler, handlerId);
        this.handlersById.set(handlerId, handler);
        return { handlerId };
    }
    getIdByHandler(handler) {
        var _a;
        return (_a = this.idsByHandler.get(handler)) !== null && _a !== void 0 ? _a : null;
    }
    getHandlerById(id) {
        var _a;
        return (_a = this.handlersById.get(id)) !== null && _a !== void 0 ? _a : null;
    }
    get() {
        return [...this.idsByHandler.entries()].map(([handler, handlerId]) => ({
            handler,
            handlerId,
        }));
    }
    delete(handler) {
        const handlerId = this.idsByHandler.get(handler);
        if (handlerId)
            this.handlersById.delete(handlerId);
        this.idsByHandler.delete(handler);
        return { isEmpty: this.idsByHandler.size < 1 };
    }
    size() {
        return this.idsByHandler.size;
    }
}
exports.SubscriptionHandlerIdMap = SubscriptionHandlerIdMap;
