// @amazon-connect/core/lib/messaging/subscription/subscription-set.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubscriptionSet = void 0;
const subscription_map_1 = require("./subscription-map");
class SubscriptionSet {
    constructor() {
        this.map = new subscription_map_1.SubscriptionMap();
    }
    add(topic, value) {
        this.map.addOrUpdate(topic, () => new Set([value]), (s) => s.add(value));
    }
    delete(topic, value) {
        const s = this.map.get(topic);
        if (s) {
            s.delete(value);
            if (s.size === 0) {
                this.map.delete(topic);
            }
        }
    }
    get(topic) {
        var _a;
        return [...((_a = this.map.get(topic)) !== null && _a !== void 0 ? _a : [])];
    }
    size(topic) {
        var _a, _b;
        return (_b = (_a = this.map.get(topic)) === null || _a === void 0 ? void 0 : _a.size) !== null && _b !== void 0 ? _b : 0;
    }
    isEmpty(topic) {
        return this.size(topic) === 0;
    }
    getAllSubscriptions() {
        return this.map.getAllSubscriptions();
    }
}
exports.SubscriptionSet = SubscriptionSet;
