// @amazon-connect/site-streams/lib/global-resiliency/global-resiliency-proxy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlobalResiliencyProxy = void 0;
const core_1 = require("@amazon-connect/core");
const global_resiliency_region_1 = require("./global-resiliency-region");
const regional_proxy_1 = require("./regional-proxy");
const verify_region_1 = require("./verify-region");
class GlobalResiliencyProxy extends core_1.Proxy {
    constructor(provider) {
        super(provider);
        this.activeRegion = global_resiliency_region_1.GlobalResiliencyRegion.Primary;
        this.regionProxies = {
            [global_resiliency_region_1.GlobalResiliencyRegion.Primary]: new regional_proxy_1.RegionalProxy({
                provider,
                region: global_resiliency_region_1.GlobalResiliencyRegion.Primary,
                getUpstreamMessageOrigin: this.getUpstreamMessageOrigin.bind(this),
                relayToGlobalResiliencyProxy: this.handleMessageFromSubject.bind(this),
            }),
            [global_resiliency_region_1.GlobalResiliencyRegion.Secondary]: new regional_proxy_1.RegionalProxy({
                provider,
                region: global_resiliency_region_1.GlobalResiliencyRegion.Secondary,
                getUpstreamMessageOrigin: this.getUpstreamMessageOrigin.bind(this),
                relayToGlobalResiliencyProxy: this.handleMessageFromSubject.bind(this),
            }),
        };
        this.proxyLogger = new core_1.ConnectLogger({
            source: "globalResiliencyProxy",
            provider,
        });
    }
    initProxy() {
        Object.values(this.regionProxies).forEach((proxy) => proxy.init());
    }
    setCCPIframe({ iframe, region }) {
        (0, verify_region_1.verifyRegion)(region);
        this.regionProxies[region].setCCPIframe(iframe);
    }
    setActiveRegion(region) {
        (0, verify_region_1.verifyRegion)(region);
        if (region !== this.activeRegion) {
            const previousRegionProxy = this.regionProxies[this.activeRegion];
            const currentRegionProxy = this.regionProxies[region];
            // Removes subscriptions from original engine
            this.unsubscribeAllHandlers();
            this.proxyLogger.info("Active region changed", {
                current: region,
                instanceUrl: currentRegionProxy.instanceUrl,
                previousInstanceUrl: previousRegionProxy.instanceUrl,
            });
            this.activeRegion = region;
            // Adds subscriptions to new engine
            this.restoreAllHandler();
            const currentStatus = this.status.getStatus();
            const activeRegionStatus = currentRegionProxy.connectionStatus;
            switch (activeRegionStatus) {
                case "ready":
                    this.proxyLogger.info("Active region is ready", {
                        activeRegionStatus,
                    });
                    this.status.update({
                        status: "ready",
                        connectionId: currentRegionProxy["connectionId"],
                    });
                    break;
                case "connecting":
                case "initializing":
                    if (currentStatus !== activeRegionStatus) {
                        this.status.update({ status: activeRegionStatus });
                    }
                    break;
                case "error":
                    if (currentStatus !== activeRegionStatus) {
                        this.status.update({
                            status: "error",
                            reason: "new active region in error on transition",
                            details: { region: this.activeRegion },
                        });
                    }
                    break;
            }
        }
    }
    get proxyType() {
        return "global-resiliency-proxy";
    }
    // Override the normal sendOrQueueMessageToSubject to rely on the
    // of the regional proxy
    sendOrQueueMessageToSubject(message) {
        this.regionProxies[this.activeRegion].sendOrQueueMessageToSubject(message);
    }
    addContextToLogger() {
        return { activeRegion: this.activeRegion };
    }
    // When sending a message, it goes to the sendOrQueueMessageToSubject of the
    // active region
    sendMessageToSubject(message) {
        this.regionProxies[this.activeRegion].sendOrQueueMessageToSubject(message);
    }
    getUpstreamMessageOrigin() {
        return Object.assign(Object.assign({ _type: "global-resiliency-streams-site", providerId: this.provider.id }, (0, core_1.getOriginAndPath)()), { activeRegion: this.activeRegion });
    }
    addChildIframeChannel(params) {
        this.regionProxies[this.activeRegion].addChildIframeChannel(params);
    }
    addChildComponentChannel(params) {
        this.regionProxies[this.activeRegion].addChildComponentChannel(params);
    }
    updateChildIframeChannelPort(params) {
        this.regionProxies[this.activeRegion].updateChildIframeChannelPort(params);
    }
}
exports.GlobalResiliencyProxy = GlobalResiliencyProxy;
