// @amazon-connect/core/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sdkVersion = exports.noResultResponseErrorKey = exports.isNoResultResponseError = exports.isClientTimeoutResponseError = exports.handlerNotFoundResponseErrorKey = exports.formatClientTimeoutError = exports.Proxy = exports.createModuleProxy = exports.resetGlobalProvider = exports.isAmazonConnectProvider = exports.getGlobalProvider = exports.AmazonConnectProviderBase = exports.SubscriptionSet = exports.SubscriptionMap = exports.isConnectError = exports.ConnectError = exports.ModuleContext = exports.Context = exports.ConnectClientWithOptionalConfig = exports.ConnectClient = void 0;
var client_1 = require("./client");
Object.defineProperty(exports, "ConnectClient", { enumerable: true, get: function () { return client_1.ConnectClient; } });
Object.defineProperty(exports, "ConnectClientWithOptionalConfig", { enumerable: true, get: function () { return client_1.ConnectClientWithOptionalConfig; } });
var context_1 = require("./context");
Object.defineProperty(exports, "Context", { enumerable: true, get: function () { return context_1.Context; } });
Object.defineProperty(exports, "ModuleContext", { enumerable: true, get: function () { return context_1.ModuleContext; } });
var error_1 = require("./error");
Object.defineProperty(exports, "ConnectError", { enumerable: true, get: function () { return error_1.ConnectError; } });
Object.defineProperty(exports, "isConnectError", { enumerable: true, get: function () { return error_1.isConnectError; } });
__exportStar(require("./logging"), exports);
var subscription_1 = require("./messaging/subscription");
Object.defineProperty(exports, "SubscriptionMap", { enumerable: true, get: function () { return subscription_1.SubscriptionMap; } });
Object.defineProperty(exports, "SubscriptionSet", { enumerable: true, get: function () { return subscription_1.SubscriptionSet; } });
__exportStar(require("./metric"), exports);
var provider_1 = require("./provider");
Object.defineProperty(exports, "AmazonConnectProviderBase", { enumerable: true, get: function () { return provider_1.AmazonConnectProviderBase; } });
Object.defineProperty(exports, "getGlobalProvider", { enumerable: true, get: function () { return provider_1.getGlobalProvider; } });
Object.defineProperty(exports, "isAmazonConnectProvider", { enumerable: true, get: function () { return provider_1.isAmazonConnectProvider; } });
Object.defineProperty(exports, "resetGlobalProvider", { enumerable: true, get: function () { return provider_1.resetGlobalProvider; } });
var proxy_1 = require("./proxy");
Object.defineProperty(exports, "createModuleProxy", { enumerable: true, get: function () { return proxy_1.createModuleProxy; } });
Object.defineProperty(exports, "Proxy", { enumerable: true, get: function () { return proxy_1.Proxy; } });
var request_1 = require("./request");
Object.defineProperty(exports, "formatClientTimeoutError", { enumerable: true, get: function () { return request_1.formatClientTimeoutError; } });
Object.defineProperty(exports, "handlerNotFoundResponseErrorKey", { enumerable: true, get: function () { return request_1.handlerNotFoundResponseErrorKey; } });
Object.defineProperty(exports, "isClientTimeoutResponseError", { enumerable: true, get: function () { return request_1.isClientTimeoutResponseError; } });
Object.defineProperty(exports, "isNoResultResponseError", { enumerable: true, get: function () { return request_1.isNoResultResponseError; } });
Object.defineProperty(exports, "noResultResponseErrorKey", { enumerable: true, get: function () { return request_1.noResultResponseErrorKey; } });
__exportStar(require("./request"), exports);
var sdk_version_1 = require("./sdk-version");
Object.defineProperty(exports, "sdkVersion", { enumerable: true, get: function () { return sdk_version_1.sdkVersion; } });
__exportStar(require("./utility"), exports);
