// @amazon-connect/core/lib/utility/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeoutTracker = exports.SubscriptionHandlerRelay = exports.getOriginAndPath = exports.generateUUID = exports.generateStringId = exports.deepClone = void 0;
var deep_clone_1 = require("./deep-clone");
Object.defineProperty(exports, "deepClone", { enumerable: true, get: function () { return deep_clone_1.deepClone; } });
__exportStar(require("./emitter"), exports);
var id_generator_1 = require("./id-generator");
Object.defineProperty(exports, "generateStringId", { enumerable: true, get: function () { return id_generator_1.generateStringId; } });
Object.defineProperty(exports, "generateUUID", { enumerable: true, get: function () { return id_generator_1.generateUUID; } });
var location_helpers_1 = require("./location-helpers");
Object.defineProperty(exports, "getOriginAndPath", { enumerable: true, get: function () { return location_helpers_1.getOriginAndPath; } });
var subscription_handler_relay_1 = require("./subscription-handler-relay");
Object.defineProperty(exports, "SubscriptionHandlerRelay", { enumerable: true, get: function () { return subscription_handler_relay_1.SubscriptionHandlerRelay; } });
var timeout_tracker_1 = require("./timeout-tracker");
Object.defineProperty(exports, "TimeoutTracker", { enumerable: true, get: function () { return timeout_tracker_1.TimeoutTracker; } });
