// @amazon-connect/site-streams/lib-esm/global-resiliency/regional-proxy.js
import { SiteProxy } from "@amazon-connect/site";
import { GlobalResiliencyRegion } from "./global-resiliency-region";
export class RegionalProxy extends SiteProxy {
    constructor({ provider, region, getUpstreamMessageOrigin, relayToGlobalResiliencyProxy, }) {
        super(provider, region === GlobalResiliencyRegion.Primary
            ? provider.config.primaryInstanceUrl
            : provider.config.secondaryInstanceUrl);
        this.getParentUpstreamMessageOrigin = getUpstreamMessageOrigin;
        this.relayToGlobalResiliencyProxy = relayToGlobalResiliencyProxy;
        this.ccpIFrame = null;
        this.region = region;
        this.unexpectedIframeWarningCount = 0;
    }
    get proxyType() {
        return "acgr-regional-proxy";
    }
    setCCPIframe(iframe) {
        const isCcpIFrameSet = Boolean(this.ccpIFrame);
        this.ccpIFrame = iframe;
        this.unexpectedIframeWarningCount = 0;
        if (isCcpIFrameSet)
            this.resetConnection("CCP IFrame Updated");
    }
    handleMessageFromSubject(msg) {
        switch (msg.type) {
            case "response":
            case "publish":
            case "error":
                this.relayToGlobalResiliencyProxy(msg);
                break;
            default:
                // All other messages handled by this proxy
                super.handleMessageFromSubject(msg);
        }
    }
    getUpstreamMessageOrigin() {
        return this.getParentUpstreamMessageOrigin();
    }
    verifyEventSource(evt) {
        const ccpIFrame = this.ccpIFrame;
        if (!ccpIFrame) {
            this.proxyLogger.error("CCP Iframe not provided to proxy. Unable to verify event to Connect to CCP.", {
                origin: evt.origin,
            });
            return false;
        }
        const valid = evt.source === ccpIFrame.contentWindow;
        if (!valid) {
            this.unexpectedIframeWarningCount++;
            if (this.unexpectedIframeWarningCount < 5) {
                this.proxyLogger.warn("Message came from unexpected iframe. Not a valid CCP. Will not connect", {
                    origin: evt.origin,
                    unexpectedIframeWarningCount: this.unexpectedIframeWarningCount,
                });
            }
        }
        return valid;
    }
    sendOrQueueMessageToSubject(message) {
        super.sendOrQueueMessageToSubject(message);
    }
    invalidInitMessageHandler() {
        // CCP sends messages via Streams
        // Take no action here
    }
}
