// @amazon-connect/core/lib-esm/messaging/subscription/subscription-manager.js
import { SubscriptionHandlerIdMap } from "./subscription-handler-id-map";
import { SubscriptionMap } from "./subscription-map";
export class SubscriptionManager {
    constructor() {
        this.subscriptions = new SubscriptionMap();
    }
    add(topic, handler) {
        return this.subscriptions
            .getOrAdd(topic, () => new SubscriptionHandlerIdMap())
            .add(handler);
    }
    get(topic) {
        var _a, _b;
        return (_b = (_a = this.subscriptions.get(topic)) === null || _a === void 0 ? void 0 : _a.get()) !== null && _b !== void 0 ? _b : [];
    }
    getById(topic, handlerId) {
        var _a, _b;
        return (_b = (_a = this.subscriptions.get(topic)) === null || _a === void 0 ? void 0 : _a.getHandlerById(handlerId)) !== null && _b !== void 0 ? _b : null;
    }
    delete(topic, handler) {
        var _a, _b;
        if ((_b = (_a = this.subscriptions.get(topic)) === null || _a === void 0 ? void 0 : _a.delete(handler).isEmpty) !== null && _b !== void 0 ? _b : false) {
            this.subscriptions.delete(topic);
        }
    }
    size(topic) {
        var _a, _b;
        return (_b = (_a = this.subscriptions.get(topic)) === null || _a === void 0 ? void 0 : _a.size()) !== null && _b !== void 0 ? _b : 0;
    }
    isEmpty(topic) {
        return this.size(topic) === 0;
    }
    getAllSubscriptions() {
        return this.subscriptions.getAllSubscriptions();
    }
    getAllSubscriptionHandlerIds() {
        return this.subscriptions
            .getAllSubscriptions()
            .reduce((acc, topic) => acc.concat(this.get(topic).map(({ handlerId }) => ({
            topic,
            handlerId,
        }))), []);
    }
}
